### ISSUE TITLE
Rpath test fails on major dev bigger 256 with assertion error

### ISSUE THREAD

**Issue by ericzolf** [rdiff-backup/rdiff-backup#101]
Trying:

```
PATH=$PWD/build/scripts-3.7:$PATH PYTHONPATH=$PWD/build/lib.linux-x86_64-3.7 python3 testing/rpathtest.py --failfast
.........Warning, making long socket did not fail
........................../usr/lib64/python3.7/unittest/case.py:628: ResourceWarning: unclosed file <_io.BufferedReader name=b'/home/elavarde/Public/rdiff-backup/build/testfiles/output/file.gz'>
  testMethod()
ResourceWarning: Enable tracemalloc to get the object allocation traceback
.testing/rpathtest.py:315: ResourceWarning: unclosed file <_io.BufferedWriter name=b'/home/elavarde/Public/rdiff-backup/build/testfiles/output/file.gz'>
  fp_out.write(s)
ResourceWarning: Enable tracemalloc to get the object allocation traceback
....F
======================================================================
FAIL: testGetnums (__main__.FilenameOps)
Test getting file numbers
----------------------------------------------------------------------
Traceback (most recent call last):
  File "testing/rpathtest.py", line 282, in testGetnums
    assert devnums == (259, 0), devnums
AssertionError: (3, 0)

----------------------------------------------------------------------
Ran 41 tests in 1.115s

FAILED (failures=1)
```

My interpretation is that there is a modulo error in the code of `src/cmodule.c` when calculating major (3 is the result of 259 modulo 256, and no other device fails like this).

It needs to be added that `testing/ctest.py` shows the same behaviour if I add `/dev/nvme0n1` to the list of files checked. At the same time, it shows that the python version of `make_file_dict` works correctly.

I see two options:

1. get rid of cmodule.c and replace with the python version for all platforms - that's probably the best option long term but I'm not sure about the performance impact.
2. fix cmodule.c - it probably means importing /usr/include/sys/sysmacros.h and/or /usr/include/bits/sysmacros.h instead of defining our own version of `major`.

Could someone with more C-knowledge have a look at the code and make a suggestion?

**Comment by ottok** [comment]
With the Dockerfile environment I am able to reproduce somewhat similar errors:
```
/build/rdiff-backup/.tox/py37/lib/python3.7/site-packages/rdiff_backup/rpath.py:1546: ResourceWarning: unclosed file <_io.BufferedWriter name=b'/build/rdiff-backup/.tox/py37/testfiles/output/rdiff-backup-data/mirror_metadata.2019-08-24T07:56:17Z.snapshot.gz'>
  if self.fileobj: return self.fileobj.close()
```

I do however not understand what is going on here. I suggest we merge #94 soon so that the barrier to entry for any new contributors would be as easy as possible, as the build and test environment is easy to spring up using the Dockerfile.

**Comment by ottok** [comment]
When I run with the exact command above I get this failure:
```
root@7acb8e87e8ef:/build/rdiff-backup# PATH=$PWD/build/scripts-3.7:$PATH PYTHONPATH=$PWD/build/lib.linux-x86_64-3.7 python3 testing/rpathtest.py --failfast
.........Warning, making long socket did not fail
......sF
======================================================================
FAIL: testCharDev (__main__.CheckTypes)
Char special files identified
----------------------------------------------------------------------
Traceback (most recent call last):
  File "testing/rpathtest.py", line 54, in testCharDev
    assert RPath(self.lc, "/dev/tty2", ()).ischardev()
AssertionError

----------------------------------------------------------------------
Ran 17 tests in 1.007s

FAILED (failures=1, skipped=1)
```

**Comment by ericzolf** [comment]
@ottok let me guess that you don't have /dev/tty2 available on your PC. This is unrelated and happens after _my_ error, it's just that this specific test makes assumptions about the running environment, we should fix this but in another issue (#108).

Regarding the original issue, I made some tests and am almost ready to push a new PR which gets rid of the C implementation (without noticeable performance issue). So stay tuned.

**Comment by ottok** [comment]
```
root@b88ff38862a7:/build/rdiff-backup# ls -lah /dev/tt* 
crw-rw-rw- 1 root root 5, 0 Aug 25 21:05 /dev/tty
```

I run there tests in Docker – therefore it should be very easy for you to reproduce the run, as environment can be downloaded and each step is documented. If you don't want Docker on your laptop, maybe run this in a cloud machine or in a virtual machine?
